httprequest_lego_provider/dns.py

```diff
@@ -0,0 +1,82 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+"""DNS utiilities."""
+
+import io
+from pathlib import Path
+from tempfile import TemporaryDirectory
+
+from git import Git, GitCommandError, Repo
+
+from .settings import DNS_REPOSITORY_URL, SSH_IDENTITY_FILE
+
+FILENAME_SUFFIX = ".domain"
+SPLIT_DNS_REPOSITORY_URL = DNS_REPOSITORY_URL.rsplit("@", 1)
+REPOSITORY_BASE_URL = SPLIT_DNS_REPOSITORY_URL[0]
+REPOSITORY_BRANCH = SPLIT_DNS_REPOSITORY_URL[1] if len(SPLIT_DNS_REPOSITORY_URL) > 1 else None
+RECORD_CONTENT = "{record} 600 IN TXT \042{value}\042\n"
+SSH_EXECUTABLE = f"ssh -i {SSH_IDENTITY_FILE}"
+
+
+class DnsSourceUpdateError(Exception):
+    """Exception for DNS update errors."""
+
+
+def write_dns_record(dns_record: str, value: str) -> None:
+    """Write a DNS record following the canonical-is-dns-configs specs if it doesn't exist.
+
+    Args:
+        dns_record: the DNS record to add.
+        value: ACME challenge for DNS record to add.
+
+    Raises:
+        DnsSourceUpdateError: if an error while updating the repository occurs.
+    """
+    with TemporaryDirectory() as tmp_dir, Git().custom_environment(GIT_SSH_COMMAND=SSH_EXECUTABLE):
+        try:
+            repo = Repo.clone_from(REPOSITORY_BASE_URL, tmp_dir, branch=REPOSITORY_BRANCH)
+            splitted_record = dns_record.rsplit(".", 2)
+            subdomain = splitted_record[0]
+            domain = ".".join(splitted_record[1:]) if splitted_record else "."
+            dns_record_file = Path(f"{repo.working_tree_dir}/{domain}{FILENAME_SUFFIX}")
+            content = dns_record_file.read_text("utf-8")
+            new_content = []
+            for line in io.StringIO(content):
+                if not line.split() or line.split()[0] != subdomain:
```

Comment:
Shouldn't we log a warning if the subdomain entry already exists and we'll override it ?

Reply:
Also I don't know how the python compiler works, but we're doing the split twice

Reply:
And just to be sure, I'd also treat differently commented lines, they are lines starting with `;`

Reply:
> Shouldn't we log a warning if the subdomain entry already exists and we'll override it ?
It doesn't seem like a special case to me. I feel like updates will probably be part of normal operation
> Also I don't know how the python compiler works, but we're doing the split twice
`split` doesn't modify the string
> And just to be sure, I'd also treat differently commented lines, they are lines starting with ;
They won't start with `subdomain`, but I can explicitly skip those if you think that's clearer

Reply:
> Also I don't know how the python compiler works, but we're doing the split twice

Doesn't feel like a big deal, this is a cheap operation, so I feel we should favour simplicity

> Shouldn't we log a warning if the subdomain entry already exists and we'll override it ?

Should we error under that scenario, perhaps?

Reply:
If we error in case of entry already there, what are the solution the user can apply to solve the situation if we're not in the race condition scenario ?
I think that's better to proceed, but log a warn